Jellyfin.Server.Implementations/Trickplay/TrickplayManager.cs

```diff
@@ -0,0 +1,468 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Jellyfin.Data.Entities;
+using MediaBrowser.Controller.Configuration;
+using MediaBrowser.Controller.Drawing;
+using MediaBrowser.Controller.Entities;
+using MediaBrowser.Controller.Library;
+using MediaBrowser.Controller.MediaEncoding;
+using MediaBrowser.Controller.Trickplay;
+using MediaBrowser.Model.Configuration;
+using MediaBrowser.Model.Entities;
+using MediaBrowser.Model.IO;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+
+namespace Jellyfin.Server.Implementations.Trickplay;
+
+/// <summary>
+/// ITrickplayManager implementation.
+/// </summary>
+public class TrickplayManager : ITrickplayManager
+{
+    private readonly ILogger<TrickplayManager> _logger;
+    private readonly IMediaEncoder _mediaEncoder;
+    private readonly IFileSystem _fileSystem;
+    private readonly EncodingHelper _encodingHelper;
+    private readonly ILibraryManager _libraryManager;
+    private readonly IServerConfigurationManager _config;
+    private readonly IImageEncoder _imageEncoder;
+    private readonly IDbContextFactory<JellyfinDbContext> _dbProvider;
+
+    private static readonly SemaphoreSlim _resourcePool = new(1, 1);
```

Comment:
I think we can allow some concurrency here.. or is this a very intensive task?

Reply:
Yeah this is a very intensive task and the intended route for improving speed is just upping the thread count and allowing ffmpeg to use multiple cores as needed which it seems to be very good at. Once a second ffmpeg process is added it just halves the speed and so on. 

We discussed the possibility of concurrency to make better use of the GPU in https://github.com/nicknsy/jellyscrub/issues/77, but even on high-end server GPUs we didn't see any improvement doing multiple ffmpeg extractions at once, though there could have been flaws in the testing methodology there.